### DISCUSSION TITLE
App feedback

### DISCUSSION THREAD

**Discussion by EarNugget** [TechbeeAT/jtxBoard#201] [Ideas]
Hello everyone, I've recently found this app and it quickly became my favorite notes app! You guys are awesome. I have feedback from my experience that I'd like to share.
- Enabling markdown view removes characters such as parenthesis and apostrophes when viewing a note
- Sometimes the last character of the note description, such as a question mark, gets deleted after having saved the note
- Sometimes an additional save prompt appears proclaiming "detected unsaved changes" after having already saved the note

I would like the ability to:
- Encrypt the notes with a password or via biometrics
- Import and export encrypted collections
- Make an alarm reoccurring (current implementation duplicates tasks)
- Filter out a tag (category) associated with notes regardless of other tags associated with it

Let me know what you guys think.

Device: Pixel 6a
OS: Graphene OS
Installation medium: F-Droid
App version: 2.1.9

**Comment by patrickunterwegs** [comment]
Hey @EarNugget , thank you for your feedback 🙂 

- Enabling markdown view removes characters such as parenthesis and apostrophes when viewing a note 
I have tried to reproduce this, but for me it worked. Can you give an example where characters are removed that are actually wanted?

- Sometimes the last character of the note description, such as a question mark, gets deleted after having saved the note 
Could you find a way to reproduce this with specific characters? I have tried it and I didn't run into such an issue.

- Sometimes an additional save prompt appears proclaiming "detected unsaved changes" after having already saved the note 
Indeed I think that's a little bug. When autosave is on and you go back (expecting the changes to get saved anyway), the message appears although it's already being saved in the background. Now covered in #202 

- Encrypt the notes with a password or via biometrics / Import and export encrypted collections
This needs probably a little bit more discussion. 
For the first part to encrypt entries within the app: What would be the main goal of the encryption? Are you thinking more like you have "Confidential" entries that should only be visible after you unlocked them? (But then maybe no encryption is needed, but a dedication filter option to only show those entries when you unlocked them through biometrics).
Or do you mean in general that all entries should be encrypted in the background (database)? (This comes with a lot of additional questions and things to consider, e.g. how should the synchronisation through DAVx5 work, how would a server handle this encryption, when you export entries encrypted, how would another application import them correctly. That also needs some investigation if there is any standard definition in order to get that done).

- Make an alarm reoccurring (current implementation duplicates tasks)
Hm, yes, I've been thinking about that. In the current implementation I left this out in order to reduce complexity. And to get it right: You imagine an alarm that e.g. reminds you every day to get a task done until you actually mark it as done, right?

- Filter out a tag (category) associated with notes regardless of other tags associated with it
I'm not sure if I understand that one. Basically you can already filter from the list by any category. If you filter by multiple categories, the logic would return all entries with ANY of the selected categories (not with all). Or am I getting something wrong here?


So I hope you're up for a little discussion on those topics 😉 


Ping @rfc2822 @devvv4ever @ArnyminerZ (if you'd also like to participate here) 😉

**Comment by EarNugget** [comment]
Thanks for your reply @patrickunterwegs. I've done some testing and have got quite the report. In addition, I've noticed a few minor annoyances since.
- Here are screenshots showing my problem with markdown. It appears parentheses, apostrophes, and quotation marks are affected. Characters such as underscore and star are also removed, but that is expected.
![Without Markdown](https://user-images.githubusercontent.com/55078900/213756991-6e6fe1d7-e145-48e0-94be-ccc40e9f79df.png)
![With Markdown](https://user-images.githubusercontent.com/55078900/213757104-ee5b6265-7086-42cc-a1c3-362d78354256.png)
- About the issue with the last character being removed, I found the problem. When making markdown bullet points, if you insert a bullet in the middle of the list it removes the last character. Will occur indefinitely.
![Character Removed](https://user-images.githubusercontent.com/55078900/213757850-4a60e264-e45b-45db-bcaa-1461672b74aa.gif)
- Regarding note encryption, I think this would add another level of "no brainer" for people to download and recommend this app. When looking for a notes app, this was originally a feature I was looking for. If all gets implemented successfully, all it takes is one fingerprint scan and one can get access to their notes. It looks like there would be a few ways to go about this (not knowing the programming side), but as long as there were an option to encrypt and unlock all of the notes, it would be good for me. I haven't used the synchronization client.
- About a reoccurring alarm, what you've suggested sounds correct. Although, for my use case I was looking to set any reoccurring duration, such as 3 months.
- About filtering out a tag, I suppose the best way to convey it is if you want to filter out all notes with the NSFW tag. If the note still has other tags associated with it, it will still show. Right now the app does work as expected, I'm just wondering if there's a way a tag can be filtered out like this. (First image, notes with filters reset. Second, trying to filter out NSFW. Third, notes with new filter.)
![1 Notes, With Filters Reset](https://user-images.githubusercontent.com/55078900/213764607-7e4fa151-4572-4792-97fc-908bf92ee6c9.png)
![2 Attempting to Filter Out NSFW](https://user-images.githubusercontent.com/55078900/213765916-b584ce4f-3d74-4c7f-8bd7-6dba1bbb5de9.png)
![3 Notes, With New Filter](https://user-images.githubusercontent.com/55078900/213764621-5a024545-c839-4600-b617-31d2b9afb608.png)

Two small annoyances since yesterday
- When I receive an alarm for a task, the timestamp of the notification sent is incorrect. For instance, I created a task yesterday at 2:30pm which only had a due date set for 8am and an alarm set "on due". When the notification popped up, it didn't say "now" it said "9h".
- Tapping on the note description when viewing a note will open up the note to be edited, but only when markdown mode is disabled. Either way, tapping on the title starts editing the note.

All of that was a mouthful, but I hope it is helpful.

**Comment by patrickunterwegs** [comment]
Hey @EarNugget , thanks for the details! I'll check them one by one. I open a dedicated thread here for this topic: 

> Regarding note encryption, I think this would add another level of "no brainer" for people to download and recommend this app. When looking for a notes app, this was originally a feature I was looking for. If all gets implemented successfully, all it takes is one fingerprint scan and one can get access to their notes. It looks like there would be a few ways to go about this (not knowing the programming side), but as long as there were an option to encrypt and unlock all of the notes, it would be good for me. I haven't used the synchronization client.

So for you it's less about the encryption, but the visibility of the entries, right? Would #205 cover what you mean?

**Reply by EarNugget** [reply]
Thank you. No, for me it is about encryption. User data in a notes app like this could be incredibly personal and I would like to remove the points of failure by being able to store the data encrypted while not in use and by being able to export and import encrypted collections directly. So as a result other apps have a setting for when the user data should be encrypted again and thus require authentication again such as when "app is no longer in focus" or "app is closed".

**Reply by patrickunterwegs** [reply]
Hm, I understand. It's a bit tricky in the context of jtx Board and I'm not sure if this can be done so soon...

If you use jtx Board only offline and only jtx Board, then okay, data can be encrypted in the database and at least the summary and description field can be encrypted in an exported ics.

But jtx Board follows the ideas of interoperability with other apps and any custom encryption would be against this idea, as the standard does not define any encryption and no other app would know how to decrypt the data...

I will continue thinking about it and also discuss it with the friends from DAVx5 if they have any good idea about this topic!

**Reply by EarNugget** [reply]
One is thankful for your efforts 😊

**Comment by patrickunterwegs** [comment]
@EarNugget  thanks for the screenshots for
> Here are screenshots showing my problem with markdown. It appears parentheses, apostrophes, and quotation marks are affected. Characters such as underscore and star are also removed, but that is expected.

I have reproduced this issue on jtx Board 2.1.9, but in 2.2.0 it doesn't occur anymore. It seems like the library update in https://github.com/ArnyminerZ/MarkdownText fixed it 🙂

**Reply by EarNugget** [reply]
Awesome!

**Comment by patrickunterwegs** [comment]
@EarNugget 
> About the issue with the last character being removed, I found the problem. When making markdown bullet points, if you insert a bullet in the middle of the list it removes the last character. Will occur indefinitely.

Thanks for the video, it's still a bit hard for me to understand what's going on there. But there was another user complaining about a similar thing. In this case it was not the app, but the keyboard provider that had a weird behaviour on the textfield, but only in jtx Board (as it might have been the only one using jetpack compose in the background). Would you mind trying if it works with an alternative keyboard? Myself I couldn't reproduce an issue like that in 2.1.9 and 2.2.0...

**Reply by EarNugget** [reply]
I'm not sure. I swapped between Gboard and my stock keyboard, no difference. Turned off and changed text correction in my device settings and couldn't produce a different result.

**Reply by patrickunterwegs** [reply]
Now I got it 🙂 it was really hard to see what was going wrong, especially as the beginning of the gif was not clear (as it was repeating). I could reproduce it now.
I'll create an issue for that!

**Reply by patrickunterwegs** [reply]
Okay, I fixed that one, a stupid small but 😉 It will be available in v2.2.1

**Reply by EarNugget** [reply]
Thanks! It will make a big difference!

**Comment by patrickunterwegs** [comment]
@EarNugget 
> About a reoccurring alarm, what you've suggested sounds correct. Although, for my use case I was looking to set any reoccurring duration, such as 3 months.

Okay, got it. Basically this is feasible, I will also refactor the alarm handling (and also revisit the last point that you mentioned) in an upcoming version. But basically may I ask you for your usecase? I mean would you have a task that just never finishes (never gets checked) but you still want to get a reminder every 3 months? In that case, wouldn't it be better to have the task itself recurring with an alarm attached?

**Reply by EarNugget** [reply]
That's a good point. It looked like the current recurrence system actually would work for my use case, but when I tried set to "ends never" it duplicated so many notes on the spot.

**Reply by patrickunterwegs** [reply]
Yes, in the background duplicates are generated, but in the list you have an option to show only one recurring entry in the future which just hides all others but the past ones and the upcoming one 😉

**Reply by EarNugget** [reply]
Oh, I missed that. So when you get an alarm you can click done, then in the tasks list you can filter out or group completed tasks where you want. Got it 😊

**Reply by patrickunterwegs** [reply]
Exactly 😉 
I'll still have a look at recurring alarms in the future, it will be tracked here: #210 . But those are more meant for example to have a daily reminder of a due task until it's marked as done!

**Comment by patrickunterwegs** [comment]
About this one: 
@EarNugget 
> About filtering out a tag, I suppose the best way to convey it is if you want to filter out all notes with the NSFW tag. If the note still has other tags associated with it, it will still show. Right now the app does work as expected, I'm just wondering if there's a way a tag can be filtered out like this. (First image, notes with filters reset. Second, trying to filter out NSFW. Third, notes with new filter.)

I'm not sure if this is such a common case. Sure something could be implemented, but at the cost of having a more complicated UI for filtering. And then you can always find even more combinations that might make sense in very specific use cases (e.g. you want to filter only entries, that have ALL selected categories assigned instead of ANY, or you want all entries with one specific category but also entries that DON'T have specific other categories...

In this case I would rather like to keep it simple and clear for everybody 😉

**Reply by EarNugget** [reply]
I agree now that the NSFW tag would be a niche scenario to deal with when filtering tags. Maybe down the line it could get implemented another way...

**Comment by EarNugget** [comment]
Another issue @patrickunterwegs I'm having with markdown that I forgot about: attempting to underline places a pair of underscores which <ins>italicizes</ins> the text. Adding another pair will bold the text. It's odd because everything else appears to work.

**Reply by patrickunterwegs** [reply]
Hey @EarNugget , thanks, indeed it's strange, either I have overlooked it while testing, or the library changed after... But anyhow, the underline tag is not even supported by the library, so I actually need to remove the underline tag. (#221)

**Comment by rogercreagh** [comment]
re tag filtering
Currently the way the filtering works is that you can select multiple tags and it will show all items with ANY of the tags.

What I have implemented elsewhere (website, not mobile) is filtering that allows selection of multiple tags and an additional selector for ANY (default), ALL, or NONE. The logic for handling this in php and mysql was a little tortuous but it works and I find it really useful. 

In jtxB I can see the use for being able to filter a journal by two (or more) tags to get only entries that are tagged with both, or filtering the list to exclude all entries tagged with a tag or both of two tags (or more) as well as the simple ANY case.
